File: python/day08.py
import heapq

class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, a, b):
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False
        if self.size[ra] < self.size[rb]:
            ra, rb = rb, ra
        self.parent[rb] = ra
        self.size[ra] += self.size[rb]
        return True

def dist2(p, q):
    dx, dy, dz = p[0]-q[0], p[1]-q[1], p[2]-q[2]
    return dx*dx + dy*dy + dz*dz

def read_points(filename):
    points = []
    with open(filename) as f:
        for line in f:
            if line.strip():
                x,y,z = map(int, line.replace(',', ' ').split())
                points.append((x,y,z))
    return points

def solve_day8(filename):
    points = read_points(filename)
    n = len(points)
    uf = UnionFind(n)

    # Build all pairs into a min-heap
    heap = []
    for i in range(n):
        for j in range(i+1, n):
            d2 = dist2(points[i], points[j])
            heapq.heappush(heap, (d2, i, j))

    part1_answer = None
    part2_answer = None
    pairs_considered = 0

    while heap:
        d2, i, j = heapq.heappop(heap)
        pairs_considered += 1
        uf.union(i, j)

        # Part One: after 1000 pairs considered
        if pairs_considered == 1000 and part1_answer is None:
            seen = {}
            sizes = []
            for idx in range(n):
                root = uf.find(idx)
                if root not in seen:
                    seen[root] = True
                    sizes.append(uf.size[root])
            sizes.sort(reverse=True)
            while len(sizes) < 3:
                sizes.append(1)
            part1_answer = sizes[0] * sizes[1] * sizes[2]

        # Part Two: when all connected
        root0 = uf.find(0)
        if uf.size[root0] == n and part2_answer is None:
            part2_answer = points[i][0] * points[j][0]
            break

    return part1_answer, part2_answer

if __name__ == "__main__":
    part1, part2 = solve_day8("input/day08.txt")
    print("Day 8 Part One Answer:", part1)
    print("Day 8 Part Two Answer:", part2)

File: perl/day08.pl
#!/usr/bin/perl
use strict;
use warnings;
use List::Util qw(max);

# -------- Union-Find --------
my @parent;
my @size;

sub uf_init {
    my ($n) = @_;
    @parent = (0 .. $n-1);
    @size   = (1) x $n;
}

sub uf_find {
    my ($x) = @_;
    $parent[$x] = uf_find($parent[$x]) if $parent[$x] != $x;
    return $parent[$x];
}

sub uf_union {
    my ($a, $b) = @_;
    $a = uf_find($a);
    $b = uf_find($b);
    return 0 if $a == $b;
    if ($size[$a] < $size[$b]) {
        ($a, $b) = ($b, $a);
    }
    $parent[$b] = $a;
    $size[$a]  += $size[$b];
    return 1;
}

sub dist2 {
    my ($p, $q) = @_;
    my ($dx,$dy,$dz) = ($p->[0]-$q->[0], $p->[1]-$q->[1], $p->[2]-$q->[2]);
    return $dx*$dx + $dy*$dy + $dz*$dz;
}

sub read_points {
    my ($filename) = @_;
    open my $fh, '<', $filename or die "Cannot open $filename: $!";
    my @pts;
    while (<$fh>) {
        chomp;
        next unless /\S/;
        my ($x,$y,$z) = split /[, ]+/, $_;
        push @pts, [int($x), int($y), int($z)];
    }
    close $fh;
    return \@pts;
}

sub main {
    my $filename = shift @ARGV || 'input/day08.txt';
    my $points   = read_points($filename);
    my $n        = @$points;

    # Build all pairs
    my @pairs;
    for my $i (0..$n-1) {
        for my $j ($i+1..$n-1) {
            push @pairs, [$i,$j,dist2($points->[$i],$points->[$j])];
        }
    }
    @pairs = sort { $a->[2] <=> $b->[2] } @pairs;

    uf_init($n);
    my $pairs_considered = 0;
    my $part1_answer;
    my $part2_answer;

    for my $pair (@pairs) {
        my ($i,$j,$d2) = @$pair;
        $pairs_considered++;
        uf_union($i,$j);   # attempt union, may or may not merge

        # Part One: after 1000 pairs considered
        if ($pairs_considered == 1000 && !defined $part1_answer) {
            my %seen;
            my @sizes;
            for my $idx (0..$n-1) {
                my $root = uf_find($idx);
                next if $seen{$root}++;
                push @sizes, $size[$root];
            }
            @sizes = sort { $b <=> $a } @sizes;
            $part1_answer = $sizes[0] * $sizes[1] * $sizes[2];
        }

        # Part Two: when all connected
        my $root0 = uf_find(0);
        if ($size[$root0] == $n && !defined $part2_answer) {
            $part2_answer = $points->[$i]->[0] * $points->[$j]->[0];
            last;
        }
    }

    print "Day 8 Part One Answer: $part1_answer\n";
    print "Day 8 Part Two Answer: $part2_answer\n";
}

main();
